src/Analytics/track.ts

```diff
@@ -107,4 +107,80 @@ export interface Track<PP = {}, SS = null, AA extends Array<any> = Array<any>> {
   ): Decorator
 }
 
+/**
+ * This is the regular `track` function, but pre-typed using our schema.
+ *
+ * For its normal usage see the docs https://github.com/NYTimes/react-tracking
+ *
+ * One important gotcha that is easily overlooked is that you always need to
+ * ‘enable’ tracking on a component before you can use it inside a component.
+ *
+ * @example
+ *
+ *     ```ts
+ *     import React from "react"
+ *     import { track } from "Analytics"
+ *
+ *     // This is what enables tracking for further use inside the component.
+ *     @track()
+ *     class Artist extends React.Component {
+ *       // This only works with the above line that enables tracking.
+ *       @track({ … })
+ *       handleClick() { … }
+ *     }
+ */
 export const track: Track = _track
+
+/**
+ * ## Writing tests for your tracked code
+ *
+ * By default we mock `react-tracking`, so it's not possible to test the code
+ * easily.
+ *
+ * A good pattern for testing analytics code is to have a completely separate
+ * file for the tests. For example: `__tests__/DateSource-analytics-tests.tsx`.
+ * Jest has each test file run in a unique environment, so in that file we can
+ * unmock `react-tracking`.
+ *
+ * Here's a full example:
+ *
+ * @example
+ *
+ *     ```ts
+ *     import { mount } from "enzyme"
+ *     import React from "react"
+ *     import { mockTracking } from "../../../../Analytics"
+ *     import { DateSource } from "../DateSource"
+ *
+ *     jest.unmock("react-tracking")
+ *
+ *     const NewsArticle = { news_source: { url: "http://nytimes.com" }}
+ *
+ *     describe("DateSource analytics", () => {
+ *       it("tracks news source link", () => {
+ *         const { Component, dispatch } = mockTracking(DateSource)
+ *         const component = mount(<Component article={NewsArticle} />)
+ *         component
+ *           .find("a")
+ *           .at(0)
+ *           .simulate("click")
+ *         expect(dispatch).toBeCalledWith({
+ *           action: "Click",
+ *           type: "external link",
+ *           label: "news source",
+ *           destination_path: "http://nytimes.com",
+ *         })
+ *       })
+ *     })
+ *     ```
+ */
+export function mockTracking<P>(
+  Subject: React.ComponentType<P>
+): {
+  Component: React.ComponentType<P>
+  dispatch: jest.Mock<(trackedData: Trackables) => void>
+} {
+  const dispatch = jest.fn() as jest.Mock<() => void>
+  const Component = _track({}, { dispatch })(Subject)
+  return { Component, dispatch }
+}
```

Comment:
This is nice, did you consider the `__mocks` file-system approach? I think I did that in Emission

Reply:
Or not, must have debated doing that a some point.

Reply:
Hmm, I’m not sure how you’d imagine that to work? In this case we’re not actually mocking and decorating the component being tested with the right context for `react-tracking` to work and capture the tracked data.

Reply:
Mentally I model the fs `__mocks` idea to be an "export API" compatible version of the mocked module that also contains introspection tools / doesn't run work outside the boundaries of that module. Then `track` would not send out to force, etc. 

This technique is different from the HOC wrapper this function actually provides though. 

IMO your technique ends up with less type faff in the test code though, because it's shipping as a part of the normal export. Example: [module](https://github.com/danger/peril/blob/master/source/db/getDB.ts) - [mock](https://github.com/danger/peril/blob/master/source/db/__mocks__/getDB.ts) -  [test faff](/Users/orta/dev/projects/danger/hazmat/peril/source/scheduler/_tests/_startScheduler.test.ts)

Reply:
I like that the types are immediately settled, but the main thing that needs done is setup that context and I don’t easily see a way to do that with a `__mocks__` mock 🤔